File: src/types/topology.ts
export type TopologyNodeKind = "project" | "commandRun" | "process" | "port" | "url";
export type TopologyEdgeKind =
  | "projectHasRun"
  | "runStartedProcess"
  | "projectContainsProcess"
  | "processParent"
  | "processOwnsPort"
  | "portExposesUrl";
export type TopologyConfidence = "certain" | "strong" | "inferred";

export interface TopologyEvidence {
  source: string;
  detail: string;
}

export interface TopologyMetadata {
  label: string;
  value: string;
}

export interface TopologyAction {
  id: string;
  label: string;
  url: string | null;
}

export interface TopologyNode {
  id: string;
  kind: TopologyNodeKind;
  label: string;
  detail: string | null;
  status: string | null;
  metadata: TopologyMetadata[];
  actions: TopologyAction[];
}

export interface TopologyEdge {
  id: string;
  from: string;
  to: string;
  kind: TopologyEdgeKind;
  confidence: TopologyConfidence;
  evidence: TopologyEvidence[];
}

export interface TopologyIssue {
  code: string;
  message: string;
}

export interface TopologyGraph {
  generatedAtMs: number;
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  issues: TopologyIssue[];
}

File: src-tauri/src/domain/topology.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TopologyNodeKind {
    Project,
    CommandRun,
    Process,
    Port,
    Url,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TopologyEdgeKind {
    ProjectHasRun,
    RunStartedProcess,
    ProjectContainsProcess,
    ProcessParent,
    ProcessOwnsPort,
    PortExposesUrl,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TopologyConfidence {
    Certain,
    Strong,
    Inferred,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyEvidence {
    pub source: String,
    pub detail: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyMetadata {
    pub label: String,
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyAction {
    pub id: String,
    pub label: String,
    pub url: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyNode {
    pub id: String,
    pub kind: TopologyNodeKind,
    pub label: String,
    pub detail: Option<String>,
    pub status: Option<String>,
    pub metadata: Vec<TopologyMetadata>,
    pub actions: Vec<TopologyAction>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyEdge {
    pub id: String,
    pub from: String,
    pub to: String,
    pub kind: TopologyEdgeKind,
    pub confidence: TopologyConfidence,
    pub evidence: Vec<TopologyEvidence>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyIssue {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TopologyGraph {
    pub generated_at_ms: u64,
    pub nodes: Vec<TopologyNode>,
    pub edges: Vec<TopologyEdge>,
    pub issues: Vec<TopologyIssue>,
}
